sdk/test/trace/probability_sampler_test.cc

```diff
@@ -0,0 +1,267 @@
+
+	sampling_result = s1.ShouldSample(&c2, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	sampling_result = s1.ShouldSample(&c3, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+
+	sampling_result = s1.ShouldSample(&c4, trace_id, "", span_kind, view);
+
+	ASSERT_EQ(Decision::RECORD_AND_SAMPLE, sampling_result.decision);
+	ASSERT_EQ(nullptr, sampling_result.attributes);
+}
+
+TEST(ProbabilitySampler, ProbabilitySamplerHalf)
+{
+	double probability = 0.001;
+	int iterations = 100000, expected_count = iterations * probability,
+		actual_count = 0, variance = 1000;
+
+	opentelemetry::sdk::trace::Sampler::SpanContext c(true, true);
+	ProbabilitySampler s(probability);
+	opentelemetry::trace::SpanKind span_kind = opentelemetry::trace::SpanKind::kInternal;
+
+	using M = std::map<std::string, int>;
+	M m1 = {{}};
+	opentelemetry::trace::KeyValueIterableView<M> view{m1};
+
+	srand(time(0));
+
+	for (int i = 0; i < iterations; ++i)
+	{
+		uint8_t buf[16];
+		for (int j = 0; j < 16; ++j)
+		{
+			buf[j] = (std::rand() % 255) + 1;
+		}
+
+		opentelemetry::trace::TraceId trace_id(buf);
+
+		uint64_t r = 0;
+		std::memcpy(&r, &trace_id, 8);
+
+		if (s.ShouldSample(&c, trace_id, "", span_kind, view).decision == Decision::RECORD_AND_SAMPLE)
+		{ ++actual_count; }
+	}
+
+	ASSERT_TRUE(actual_count < (expected_count + variance));
+	ASSERT_TRUE(actual_count > (expected_count - variance));
+}
+
+TEST(ProbabilitySampler, ProbabilitySamplerOnePercent)
+{
+	double probability = 0.01;
+	int iterations = 100000, expected_count = iterations * probability,
+		actual_count = 0, variance = 1000;
+
+	opentelemetry::sdk::trace::Sampler::SpanContext c(true, true);
+	ProbabilitySampler s(probability);
+	opentelemetry::trace::SpanKind span_kind = opentelemetry::trace::SpanKind::kInternal;
+
+	using M = std::map<std::string, int>;
+	M m1 = {{}};
+	opentelemetry::trace::KeyValueIterableView<M> view{m1};
+
+	srand(time(0));
+
+	for (int i = 0; i < iterations; ++i)
+	{
+		uint8_t buf[16];
+		for (int j = 0; j < 16; ++j)
+		{
+			buf[j] = (std::rand() % 255) + 1;
+		}
+
+		opentelemetry::trace::TraceId trace_id(buf);
+
+		uint64_t r = 0;
+		std::memcpy(&r, &trace_id, 8);
+
+		if (s.ShouldSample(&c, trace_id, "", span_kind, view).decision == Decision::RECORD_AND_SAMPLE)
+		{ ++actual_count; }
+	}
+
+	ASSERT_TRUE(actual_count < (expected_count + variance));
+	ASSERT_TRUE(actual_count > (expected_count - variance));
+}
+
+TEST(ProbabilitySampler, ProbabilitySamplerAll)
+{
+	double probability = 1.0;
+	int iterations = 100000, expected_count = iterations * probability,
+		actual_count = 0;
+
+	opentelemetry::sdk::trace::Sampler::SpanContext c(true, true);
+	ProbabilitySampler s(probability);
+	opentelemetry::trace::SpanKind span_kind = opentelemetry::trace::SpanKind::kInternal;
```

Comment:
We could factor out the code below into a separate function and use this in other test cases to avoid duplicate code. Like:
```c
int RunShouldSampleCountDecision(SpanContext& context, ProbabilitySampler& sampler, int iterations, Decision decision);
```

Then test code could be shortened for each test:
```c
actual_count = RunShouldSampleCountDecision(context, sampler, iterations, Decision::RECORD_AND_SAMPLE);
ASSERT_TRUE(actual_count < (expected_count + variance));
ASSERT_TRUE(actual_count > (expected_count - variance));
```

Reply:
Good idea. I've got that changed.